pkg/recommendation/device.go

```diff
@@ -0,0 +1,235 @@
+package recommendation
+
+import (
+	"sort"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	"mayadata.io/cstorpoolauto/types"
+	bdutil "mayadata.io/cstorpoolauto/util/blockdevice"
+)
+
+// nodeCapacityBlockDevice contains key with node name and value with capacityBlockDevice
+type nodeCapacityBlockDevice map[string]capacityBlockDevices
+
+// capacityBlockDevices contains a key with capacity of block device and
+// value with all the block devices of that capacity.
+type capacityBlockDevices map[int64][]bdutil.MetaInfo
+
+type cStorPoolClusterRecommendationRequest types.CStorPoolClusterRecommendationRequest
+
+// NewDeviceRequest returns a device request object after validation.
+func NewDeviceRequest(
+	raidConfig *types.RaidGroupConfig, blockDeviceList *unstructured.UnstructuredList, poolCapacity *resource.Quantity) (
+	*cStorPoolClusterRecommendationRequest, error) {
+
+	if raidConfig == nil {
+		return nil, errors.New(
+			"Unable to create device recommendation request: Got nil raid config")
+	}
+	if blockDeviceList == nil {
+		return nil, errors.New(
+			"Unable to create device recommendation request: Got nil block device list")
+	}
+
+	err := raidConfig.Validate()
+	if err != nil {
+		return nil, errors.Wrap(err, "Unable to create capacity recommendation request")
+	}
+
+	return &cStorPoolClusterRecommendationRequest{
+		Spec: types.CStorPoolClusterRecommendationRequestSpec{
+			PoolCapacity:    *poolCapacity,
+			BlockDeviceList: *blockDeviceList,
+			DataConfig:      *raidConfig,
+		},
+	}, nil
+}
+
+// GetRecommendation returns recommended block devices for all nodes and device types.
+func (r cStorPoolClusterRecommendationRequest) GetRecommendation() map[string]types.CStorPoolClusterRecommendation {
+
+	cStorPoolClusterRecommendation := make(map[string]types.CStorPoolClusterRecommendation)
+
+	if len(r.Spec.BlockDeviceList.Items) == 0 {
+		return cStorPoolClusterRecommendation
+	}
+
+	if err := r.Spec.DataConfig.Validate(); err != nil {
+		return cStorPoolClusterRecommendation
+	}
+
+	availableBlockDeviceList := unstructured.UnstructuredList{}
+	availableBlockDeviceList.Object = r.Spec.BlockDeviceList.Object
+	for _, bd := range r.Spec.BlockDeviceList.Items {
+		isEligible, err := bdutil.IsEligibleForCStorPool(bd)
+		if err == nil && isEligible {
+			availableBlockDeviceList.Items = append(availableBlockDeviceList.Items, bd)
+		}
+	}
+	if len(availableBlockDeviceList.Items) == 0 {
+		return cStorPoolClusterRecommendation
+	}
+
+	deviceTypeNodeBlockDeviceMap := bdutil.GetTopologyMapGroupByDeviceTypeAndBlockSize(availableBlockDeviceList)
+	if len(deviceTypeNodeBlockDeviceMap) == 0 {
+		return cStorPoolClusterRecommendation
+	}
+
+	for kind, nodeBlockDeviceListMap := range deviceTypeNodeBlockDeviceMap {
```

Comment:
I have my own reservations in this logic. However as discussed earlier, we shall tackle this later.
However, if we start coding in this patter (this has become a pattern), we might find it very difficult to refactor given the time & resource constraints.

Reply:
Yes, @AmitKumarDas refactoring of this part is in my mind. I will set up one call for this. Maybe that clears some doubts and you also can provide a better way to it.
Actually what we are trying to achieve by creating topology is that - Decoupled main business logic from topology creation.
ie - For now, we are creating topology using device type (HDD or SSD) and Block size of the device. Now just think we are able to build a configurable topology by taking user's input let's say the user wants to have the same vendor in a recommendation then we need to build a topology with Vendor key like - `SSD-WD-4096` or `SSD-samsung-4096` and we will be done no need to change main business logic.
Hope I am able to give some details.
and the only change I can see in this file is - https://github.com/mayadata-io/cstorpoolauto/pull/97/files#diff-f79a4dcc36e408e0a9ef25c006a9472bR76

Reply:
I am up for discussion on this:raised_hands:

Reply:
Maybe we can have some topology building inputs like -
```
1. shouldConsiderVendor (true/false) and there will be a default value
2. shouldConsiderPhysicalBlockSize (true/false) and there will be a default value
3. shouldConsiderLogicalBlockSize (true/false) and there will be a default value
4. shouldConsiderDriveType (true/false) and there will be a default value
6. shouldConsiderDeviceType (true/false) and there will be a default value

Reply:
Including @avishnu  here.

Reply:
IMO we can have it in `V2` This may require some minor(IMO) design(UI) changes also.
Please provide your valuable thoughts here.